client/services/ReinforcementGeneratorService.js

```diff
@@ -0,0 +1,99 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for generating positive reinforcement when a user
+ * submits code based on the tasks that the code passes/fails on.
+ */
+
+tie.factory('ReinforcementGeneratorService', [
+  function() {
+
+    return {
+      getReinforcement: function(task, codeEvalResult) {
+
+        // Initializing question reinforcement data if not done already
+        if (!task.passedList) {
+          task.passedList = {};
+        }
+        if (!task.pastFailsList) {
```

Comment:
I think I understand why you're doing it like this -- you want to maintain history. However, defining stuff on the task dict midstream is not the right way to do this, since the task is supposed to be independent of the individual student's playthrough.

So, there are two ways we can go here:
(a) The "proper" way to do this is probably to preserve info about "past fails" somewhere. Perhaps you can store these in the transcript, and extract them from there.
(b) If you want to get this PR in quicker, though, that may be too much to do. In that case, let's drop the "past fails" requirement for this PR, and handle things in a stateless manner.

I'm happy either way. Let me know if you want to discuss any of this.